### DISCUSSION TITLE
support for multi-tenancy

### DISCUSSION THREAD

**Discussion by joelrealubit** [serverpod/serverpod#338] [General]
Hi. Maybe it's too early to ask, but is there any plan to support multi-tenancy? maybe either with schema-based or with a separate db per tenant?

TIA

**Comment by vlidholt** [comment]
Can you describe your use case a bit more? Serverpod currently just support one database, but you can potentially fire up multiple servers that connect to different databases on different sub-domains.

**Reply by joelrealubit** [reply]
i guess a typical use case would be in an amazon-style online store kind of system, where different vendors or shops could set up storefronts on the app. Each vendor would maybe need to have their data separate from other vendors'. The consumers would have a single login on the app but is able to transact with any store on the app. Meanwhile, store admins could have their own dashboards and admin pages where they upload their products and set up prices and promotions and such at the backend. Maybe this multi-tenancy can be done where all the vendors are in one database and one schema, with each vendor having their unique vendor ID or tenant ID, but it could also be done with all vendors are in one database but each would have a different schema, or, each vendor would just straight up have their own separate database. There are pros and cons to each approach to multi-tenancy but it would be nice if the dev can have the option to choose which type of multi-tenancy works for their context, and to have a nice dev experience using support for multi-tenancy that's built into the serverpod framework.  

This idea can also work not just with storefronts, but with any SAAS that has different companies as customers using it -- each company would need to have their own data isolated from everybody else's. Any kind of accounting system, inventory system, or workforce scheduling system, etc, if it's being done as a SAAS where companies could sign up to use it, it'll need multi-tenancy in the backend, I imagine.

**Reply by michaelmworthington** [reply]
I am just getting started with Serverpod, looking to set up a utility application to provide a front end to query our multi-tenant database.

The tenants in our database are segregated by schema. What is the recommended approach to having Serverpod query the same models across schemas?

For example, `select * from schemaone.orders` and `select * from schematwo.orders`.

I would be open to making either one or two API calls from the client. At the moment, I am working around this by creating a view in the `public` schema that does a union to return all the data. It would be nice if I had some control from within the server to selectively choose the schema to pull the data.

**Reply by vlidholt** [reply]
Currently, Serverpod only supports one schema (public). You can either set up multiple tables, or use an identifier in table column. Another option can be to have a complete setup per client.

Often, there are better solutions. So it may be smarter to look at what you are trying to achieve? Can you describe on a higher level what you want to build?

**Reply by michaelmworthington** [reply]
Thank you for the background @vlidholt, that helps point me in a direction.

At a high level, we have a web application built around `django-tenants`. I am looking to build a UI to support our ETL application that works with the same database models. However, the ETL parts are not python and do not have their own API. Like you said, it might be better to create the backend using a parallel `django-tenants` micro-service, but I was hoping to keep all of the ETL parts in dart.

**Reply by BDZK** [reply]
Is it possible to alter the session Object in every endpoint function so that it may pick a different config and password files on the fly?

**Comment by BDZK** [comment]
@vlidholt  i am really interested in your product. But multi tenancy is a critical feature for me. How can i extend a model to use different tables per client dynamically? Is there a way to manage a prefix for tables in the models so that the prefix should be added dynamically on each query?

**Comment by macMikey** [comment]
@BDZK i've been thinking about this, as well. what about setting up a server-only tenant table, and in the server endpoints calling a method that would limit fetches to those that match the user's tenant id?  how else would you imagine implementing this if you are not going to spawn a separate instance per tenant?
ultimately, it might be nice to have a scope that is related to the tenant, but that's kind-of what i'm envisioning doing, above, manually..

**Reply by BDZK** [reply]
Using a single db for all users is not my preferred solution. Because all data all mixed together. And if a user is asking for a backup of its db it is tedious to do it. Postgres schema is better in my opinion than using a filtering scope on table. Another good option is to use a prefix to all tables belonging to a given user.

**Comment by iampopal** [comment]
We also provide multiple databases to each user. creating multiple databases in google cloud SQL is okay for us todo. we need a way if possible so every user insert their data on their own database. because every company we work with want's a separated database.

**Reply by BDZK** [reply]
I prefer this approach, a db per user. That would be the perfect solution. So every user is concealed to its own db. But that means in every request, the connection to the tenant db is switched according to a tenant id stored in a token of the request...

**Reply by macMikey** [reply]
@bdzk can you filter by url? then you embed the entity in the url.

**Comment by macMikey** [comment]
i completely agree that multiple instances is the simplest/best(?) option. i'm just imagining the scenarios where one has ten, fifteen, or twenty customers, with lots of tables, but not a lot of data in any instance, and the overhead involved with supporting that sort of setup.
i've had to confront that with one of our (not flutter/serverpod) apps.

**Reply by iampopal** [reply]
Exactly

**Comment by iampopal** [comment]
How about if a userOfId: 123 request we set another config by code in a
middle ware of something so then we will not have a database specific code
in every endpoint but just in middleware which will check userId and set
that user specific database.

And yes downloading that user specific database make it much easier and
increases trust level of clients as they can download their database
anytime.
